SIMSBIOHUB-653: Observation CSV Import

MacDeluca opened the pull request:
## Links to Jira Tickets

- https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-653

## Description of Changes

- Updates existing Observation CSV import to use the new structure (CSVConfig)
- Added taxon row validator - that accepts scientific name OR TSN
- Deprecated `process` endpoint and related S3 pre-upload/process code
- Refactored existing logic to handle CSVErrors
- Created new validation functions for environments / measurements
- Added row-state getters which validate the row state to ensure correct end state
- Updated measurement service to use new row-state getters
- Added `CaseInsensitiveMap` data strucuture class
- Updated Observation `upload` endpoint -> `import`
- Modified frontend useObservationApi hook slightly and included `surveySampleSiteId`

## Testing Notes

- Should be able to import this file with a matching `Technique` and period
[Bulk Observations.csv](https://github.com/user-attachments/files/18607610/Bulk.Observations.csv)

## Notes:
- I am chipping away slowly on test coverage

MacDeluca committed:
feat: created import observation service boilerplate

MacDeluca committed:
feat: moved observation files

MacDeluca committed:
feat: starting dynamic header validator

MacDeluca committed:
Merge branch 'dev' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-653-observations-csv

MacDeluca committed:
feat: added new method for searching for taxons by scientific name

MacDeluca committed:
feat: added caseinsensitivemap + updated observation dynamic header configs

MacDeluca committed:
feat: moved files around + created getters for state values

MacDeluca committed:
feat: refactored existing getCSVConfig into smaller modular methods

MacDeluca committed:
feat: added environment utils and validators

MacDeluca committed:
feat: updated dynamic measurement validator to use shared qualitative / quantitative validators

MacDeluca committed:
feat: added type guards for environment and helper for getting tsns from TaxonMap

MacDeluca committed:
doc: updated documetation and signatures for typeguard related functions

MacDeluca committed:
feat: moved logic into own function in observation-service

MacDeluca committed:
feat: added row state getters for measurements

MacDeluca committed:
fix: changed qualitative / quantitative function signatures

MacDeluca committed:
feat: api compiling

MacDeluca committed:
feat: standard columns working

MacDeluca committed:
doc: updated function documentation

MacDeluca committed:
chore: removed deprecated test files + adding tests for simple functions

MacDeluca committed:
chore: updated tests for utils

MacDeluca committed:
feat: updated jsdoc + added tests for case-insensitive-map

MacDeluca committed:
fix: merge conflict

MacDeluca committed:
fix: dropped .only

MacDeluca committed:
chore: taxon-row-validator tests

MacDeluca committed:
chore: platform-service tests

MacDeluca committed:
fix: broken app test

NickPhura commented:
When importing an observation file, some things:

If I set the `comment` cell to a number:
![image](https://github.com/user-attachments/assets/08bce679-a6a4-4975-aef4-50ef5834c0d6)

When I have no measurements or environment columns:
![image](https://github.com/user-attachments/assets/b6a45285-8d41-49ed-9adf-2346629c9cda)

MacDeluca committed:
fix: description comment field now allows numbers

MacDeluca committed:
fix: issue with empty / trimmable headers

NickPhura commented:
Weird edge case where I've set the date/time to be the exact start date/time.

![image](https://github.com/user-attachments/assets/2b2cc36c-6c95-4b10-8a55-67e425723159)

![image](https://github.com/user-attachments/assets/178e36bc-4b7a-4dc5-90eb-258c1a7a1485)

NickPhura committed:
Fix weird milliseconds rounding issue

MacDeluca committed:
feat: fixed small issue with descriptions

MacDeluca commented:
> When importing an observation file, some things:
> 
> If I set the `comment` cell to a number:

This should be resolved now, updated the description validator to coerce numbers into strings
> When I have no measurements or environment columns:

Updated the worksheet utils to handle this.

MacDeluca committed:
chore: added tests for importObservationsService

MacDeluca committed:
Merge branch 'SIMSBIOHUB-653-observations-csv' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-653-observations-csv

MacDeluca committed:
feat: added tests for measurement-dynamic-headers-config

MacDeluca committed:
feat: added tests for dynamic measurement / environment validators

MacDeluca committed:
feat: modified row validators and created new location row validator

MacDeluca committed:
feat: added missing tests for sampling row validator

MacDeluca committed:
chore: added tests for location row validator

NickPhura commented on api/src/services/import-services/measurement/utils/measurement-utils.ts:
```diff
@@ -0,0 +1,22 @@
+ * @returns {*} {(params: CSVParams) => number} The get measurement row TSN callback
+ */
+export const getTsnFromMeasurementRow = (
+  surveyAliasMap: Map<string, ICritterDetailed>,
```
super nit: The name of this param might be more clear as something like "critterAliasMap" or "surveyCritterAliasMap"

Same for the calling function, which also names the variable "surveyAliasMap".

MacDeluca replied:
Yea that's totally fair. I just checked and there's about 90 references, I'll probably just follow this PR with one that find and replaces those references.

NickPhura replied:
Yeah fair. Not actually a big deal, just caught my eye and I was thinking "surveys have aliases?"

MacDeluca replied:
Wrote a note and will resolve in the next PR

NickPhura commented on api/src/services/import-services/observation/utils/observation-sampling-row-validator.ts:
```diff
@@ -0,0 +1,446 @@
+    const worksheetHasSamplingInformation = worksheetSiteName || worksheetTechniqueName || worksheetPeriod;
+
+    // VALID: No sampling information provided, but observation date / time is provided (no period needed)
+    if (!worksheetHasSamplingInformation && worksheetObservationDate) {
+      return [];
+    }
+
+    // INVALID: No sampling information or observation date / time provided
+    if (!worksheetHasSamplingInformation && !worksheetObservationDate) {
+      return [
+        {
+          error: 'Row does not contain sampling information or observation date / time',
+          solution: 'Please provide sampling information or a valid observation date and time',
+          header: null,
+          cell: null
+        }
+      ];
+    }
```
Does this need to check for lat/long as well?

MacDeluca replied:
I actually handle this case in `observation-location-row-validator.ts`

NickPhura replied:
I wonder if the error messages from this validator and the lat/lng validator should be more aligned, just because they are doing very similar checks regarding defined vs undefined sampling.

`Latitude is required when sampling information is not provided`
`Please provide sampling information or a valid latitude`

`Longitude is required when sampling information is not provided`
`Please provide sampling information or a valid longitude`

`Date / time is required when sampling information is not provided`
`Please provide sampling information or a valid observation date / time`

MacDeluca replied:
Yea I was uncertain if it was better to have this all combined or not. Mostly have them seperate as it was an afterthought tbh. I can combine them and drop the other validator.

NickPhura replied:
I dont think separate validators is necessarily an issue. I mostly just mean having the error messages written in a similar way, whether its across multiple validators or not.

MacDeluca committed:
fix: slight tweak to location row validator

MacDeluca committed:
fix: using getWorksheetHeader for location row validator

NickPhura commented on api/src/services/import-services/utils/qualitative.ts:
```diff
@@ -0,0 +1,67 @@
+
+  const matchingQualitativeOption = qualitativeTypeDefinition.options.find(
+    // Not sure why I need to cast params.value to string here after checking above...
+    (option) => option.option_name.toLowerCase() === String(value).toLowerCase()
```
My vscode seems happy if I remove the String cast. Seems to know it is a string.

NickPhura replied:
The similar comment in `measurement-header-configs.ts` is because its a value from the params object. If you assigned it to a new variable, and then checked the type, it would be happy.

Since the object is passed by reference, the function cant be certain that the property hasn't changed between when it was type checked and when it was used in the find.

MacDeluca replied:
Oh I just realized I forgot to drop that file `measurement-header-configs.ts` (it's deprected now and replaced with `measurement-dynamic-header-config.ts`)

MacDeluca replied:
Interesting, my editor also knows it's a string now, I must be going crazy

NickPhura replied:
haha

MacDeluca committed:
fix: dropped deprecated file

NickPhura commented on api/src/services/import-services/utils/quantitative.ts:
```diff
@@ -0,0 +1,66 @@
+  }
+
+  // Validate value is withing the environment min max bounds
+  if (quantitativeTypeDefinition.max != null && value > quantitativeTypeDefinition.max) {
```
Should this be `>=` rather than `>`?

Same for the min?

MacDeluca committed:
fix: dropped String cast

NickPhura commented on api/src/services/import-services/utils/row-state.ts:
```diff
@@ -0,0 +1,97 @@
+  return (rowOrState: CSVRow | CSVRow[typeof CSVRowState]): z.infer<SchemaType> => {
+    // Check if the state or row is being passed
+    // Note: The row state is stored as a symbol on the row object
+    const isRow = Object.getOwnPropertySymbols(rowOrState).includes(CSVRowState);
+
+    // Get the state from the row
+    const state = isRow ? rowOrState?.[CSVRowState] : rowOrState;
+
```
(optional) I think this type guard works (just to resolve the complaint about `rowOrState?.[CSVRowState]`).

```
const isRow = isCSVRow(rowOrState);
const state = isRow ? rowOrState[CSVRowState] : rowOrState;

export const isCSVRow = (rowOrState: CSVRow | CSVRow[typeof CSVRowState]): rowOrState is CSVRow => {
  return Object.getOwnPropertySymbols(rowOrState).includes(CSVRowState);
};
```

MacDeluca replied:
This is another one of those cases where my editor isn't picking up on these warnings, so my apologies. (I still need to investigate why this is happening)

Fixed.

MacDeluca replied:
Just let me know on your end if it is still present

NickPhura replied:
Seems happy on my end.

NickPhura commented on app/src/features/surveys/observations/sampling-sites/components/ImportObservationsButton.tsx:
Should this file live higher up? Since it is used by the sampling component but also by the parent table component.

MacDeluca replied:
Yea that would probably make sense, I think there was two version of this file and I probably deleted the wrong one (that was higher up) incorrectly.

MacDeluca replied:
updated

NickPhura replied:
Nit: can we put it inside a `components` folder at its current level. Just so its not sitting alongside the `SurveyObservationPage.tsx` component, which is really the entrypoint for that whole observations feature.

Actually, the `SurveyObservationHeader.tsx` file could also be moved to that components folder, since its just a child of the page component.

MacDeluca replied:
Yea I can do that

MacDeluca replied:
Fixed

NickPhura commented on api/src/utils/xlsx-utils/worksheet-utils.ts:
```diff
@@ -149,6 +149,11 @@ export const getWorksheetRowObjects = (worksheet: xlsx.WorkSheet): Record<symbol
     let rowHasValues = false;
 
     for (let j = 0; j <= originalRange.e.c; j++) {
+      // Skip empty headers
+      if (!headers[j]) {
+        continue;
+      }
+
       // Always add the header (key) to the row object
       rowObject[headers[j]] = undefined;
```
Just checking: 

Does this skip empty headers check contradict the comment below about always adding the header key?

(As to whether its best to omit empty headers or not, I can't say)

MacDeluca replied:
Yea they do contradict a bit, nice catch

Updated

NickPhura commented on api/src/utils/xlsx-utils/cell-utils.ts:
```diff
@@ -56,19 +49,22 @@ export function replaceCellDates(cell: CellObject): CellObject {
-
+    // Round the time fraction to the nearest millisecond to avoid floating point errors
+    // Excel stores time as a fraction of a day, where each day has 86400000 milliseconds.
+    // Ex: , 0.25 is 6:00:00 AM, 0.5 is 12:00:00 PM, 0.75 is 6:00:00 PM
```
👀 Oh gawd that random comma 😄

MacDeluca replied:
rm -rf *

NickPhura commented on api/src/services/import-services/observation/utils/observation-location-row-validator.ts:
```diff
@@ -0,0 +1,53 @@
+
+    const errors: CSVRowError[] = [];
+
+    if (!worksheetLatitude) {
+      errors.push({
+        error: 'Latitude is required when sampling information is not provided',
+        solution: 'Please include a latitude for the observation',
+        header: utils.getWorksheetHeader('LATITUDE', params.row),
+        cell: null
+      });
+    }
+
+    if (!worksheetLongitude) {
+      errors.push({
+        error: 'Longitude is required when sampling information is not provided',
+        solution: 'Please include a longitude for the observation',
+        header: utils.getWorksheetHeader('LONGITUDE', params.row),
+        cell: null
+      });
+    }
```
We could technically have a third error message for when BOTH latitude and longitude are undefined.

Could be as combined message similar to the `date / time` message here: https://github.com/bcgov/biohubbc/pull/1474#discussion_r1941782257

`Latitude / longitude is required when sampling information is not provided` etc

MacDeluca replied:
I have them seperate so the `header` value is more clear about what it is referencing. It will return two errors when both are missing.

MacDeluca replied:
I might be missunderstanding though

MacDeluca replied:
Outdated. Dropped this and included this logic in the sampling site validator.

NickPhura replied:
> I have them seperate so the header value is more clear about what it is referencing. It will return two errors when both are missing.

Right, was forgetting about the header field. Ignore me!

MacDeluca committed:
fix: PR comments

MacDeluca committed:
fix: COMMA DIE

MacDeluca committed:
fix: updated tests and sampling information row validator

MacDeluca committed:
fix: slightly tweaked row-state logic

MacDeluca committed:
chore: moved components into specific directory

NickPhura commented:
A few things:

If I add a measurement column to the csv that isnt supported by the taxon, the error is just the generic dynamic header error (rather than something more specific to the measurement not being valid or found or something).

If I add an empty column to my file, all of the columns after it get ignored.

MacDeluca committed:
fix: small tweak to make qualitative / quantitative validation more specific

MacDeluca committed:
Merge branch 'dev' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-653-observations-csv

MacDeluca committed:
fix: generic error dialog snackbars

NickPhura reviewed (approved):
🔍

MacDeluca committed:
Merge branch 'dev' into SIMSBIOHUB-653-observations-csv
